applications/FluidDynamicsApplication/custom_processes/shock_capturing_entropy_viscosity_process.cpp

```diff
@@ -0,0 +1,390 @@
+
+    bool missing_entropy;
+    bool missing_density;
+    bool missing_pressure;
+    bool missing_temperature;
+    bool missing_velocity;
+
+    std::tie(missing_entropy, missing_density, missing_pressure, missing_temperature, missing_velocity) =
+    block_for_each<MultiOrReduction>(mrModelPart.Nodes(), [](const NodeType& r_node)
+    {
+        const bool missing_entropy = ! r_node.SolutionStepsDataHas(NUMERICAL_ENTROPY);
+        const bool missing_density = ! r_node.SolutionStepsDataHas(DENSITY);
+        const bool missing_pressure = ! r_node.SolutionStepsDataHas(PRESSURE);
+        const bool missing_temperature = ! r_node.SolutionStepsDataHas(TEMPERATURE);
+        const bool missing_velocity = ! r_node.SolutionStepsDataHas(VELOCITY);
+
+        return std::tie(
+            missing_entropy, missing_density, missing_pressure, missing_temperature, missing_velocity
+        );
+    });
+
+    KRATOS_ERROR_IF(missing_entropy) << "Missing NUMERICAL_ENTROPY variable from one or more nodes" << std::endl;
+    KRATOS_ERROR_IF(missing_density) << "Missing DENSITY variable from one or more nodes" << std::endl;
+    KRATOS_ERROR_IF(missing_pressure) << "Missing PRESSURE variable from one or more nodes" << std::endl;
+    KRATOS_ERROR_IF(missing_temperature) << "Missing TEMPERATURE variable from one or more nodes" << std::endl;
+    KRATOS_ERROR_IF(missing_velocity) << "Missing VELOCITY variable from one or more nodes" << std::endl;
+
+    if(mrModelPart.ElementsBegin() != mrModelPart.ElementsEnd())
+    {
+        KRATOS_ERROR_IF_NOT(mrModelPart.ElementsBegin()->GetProperties().Has(HEAT_CAPACITY_RATIO))
+            << "Variable HEAT_CAPACITY_RATIO missing from elemental properties." << std::endl;
+
+#ifdef KRATOS_DEBUG
+        const double first_gamma = mrModelPart.ElementsBegin()->GetProperties().GetValue(HEAT_CAPACITY_RATIO);
+
+        const bool non_uniform_gamma =
+        block_for_each<OrReduction>(mrModelPart.Elements(), [&first_gamma](Element& r_element)
+        {
+            static constexpr double tolerance = 1e-8;
+            const double this_gamma = r_element.GetProperties().GetValue(HEAT_CAPACITY_RATIO);
+            return std::fabs(first_gamma - this_gamma) > tolerance;
+        });
+
+        KRATOS_ERROR_IF(non_uniform_gamma)
+            << "HEAT_CAPACITY_RATIO is not constant in the domain" << std::endl;
+#endif
+    }
+
+    return err_code;
+}
+
+
+const Parameters ShockCapturingEntropyViscosityProcess::GetDefaultParameters() const
+{
+    return Parameters(R"(
+    {
+        "model_part_name" : "",
+        "calculate_nodal_area_at_each_step" : false,
+        "entropy_constant"  : 0.0,
+        "energy_constant"   : 0.0,
+        "artificial_mass_viscosity_Prandtl"   : 0.1,
+        "artificial_conductivity_Prandtl"     : 0.1
+    }
+    )");
+}
+
+
+double ShockCapturingEntropyViscosityProcess::TotalDerivativeUtil::
+Divergence(const Matrix& rShapeFunGradients, const Matrix& rNodalValues)
```

Comment:
```suggestion
Divergence(
    const Matrix& rShapeFunGradients,
    const Matrix& rNodalValues)
```
(minor style guide suggestion)